nodebuilder/store.go

```diff
@@ -182,3 +182,61 @@ func indexPath(base string) string {
 func dataPath(base string) string {
 	return filepath.Join(base, "data")
 }
+
+// constraintBadgerConfig returns BadgerDB configuration optimized for low memory usage and more frequent
+// compaction which prevents memory spikes.
+// This is particularly important for LNs with restricted memory resources.
+//
+// With the following configuration, a LN uses up to 300iB of RAM during initial sync/sampling
+// and up to 200MiB during normal operation. (on 4 core CPU, 8GiB RAM droplet)
+//
+// With the following configuration and "-tags=jemalloc", a LN uses no more than 180MiB during initial
+// sync/sampling and up to 100MiB during normal operation. (same hardware spec)
+// NOTE: To enable jemalloc, build celestia-node with "-tags=jemalloc" flag, which configures Badger to
+// use jemalloc instead of Go's default allocator.
+//
+// TODO(@Wondertan): Consider alternative less constraint configuration for FN/BN
+// TODO(@Wondertan): Consider dynamic memory allocation based on available RAM
+func constraintBadgerConfig() *dsbadger.Options {
+	opts := dsbadger.DefaultOptions // this must be copied
+	// ValueLog:
+	// 2mib default => 500b - makes sure headers and samples are stored in value log
+	// This *tremendously* reduces the amount of memory used by the node, up to 10 times less during
+	// compaction
+	opts.ValueThreshold = 500
+	// make sure we don't have any limits for stored headers
+	opts.ValueLogMaxEntries = 100000000
+	// run value log GC more often to spread the work over time
+	opts.GcInterval = time.Minute * 1
+	// default 0.5 => 0.125 - makes sure value log GC is more aggressive on reclaiming disk space
+	opts.GcDiscardRatio = 0.125
+	// Snappy saves us ~200MiB of disk space on the mainnet chain to the commit's date
+	// TODO(@Wondertan): Does it worth the overhead?
+	opts.Compression = options.Snappy
+	opts.BlockCacheSize = 16 << 20
+
+	// MemTables:
+	// default 64mib => 16mib - decreases memory usage and makes compaction more often
+	opts.MemTableSize = 16 << 20
+	// default 5 => 3
+	opts.NumMemtables = 3
+	// default 5 => 3
+	opts.NumLevelZeroTables = 3
+	// default 15 => 5 - this prevents memory growth on CPU constraint systems by blocking all writers
+	opts.NumLevelZeroTablesStall = 5
+
+	// Compaction:
+	// Dynamic compactor allocation
+	compactors := runtime.NumCPU() / 2
+	if compactors < 2 {
+		compactors = 2 // can't be less than 2
+	}
+	if compactors > 7 { // ensure there is no more compactors than db table levels
+		compactors = 7
+	}
+	opts.NumCompactors = compactors
+	// makes sure badger is always compacted on shutdown
+	opts.CompactL0OnClose = true
```

Comment:
When we offload the shares and headers to the value log, the compaction becomes snappy. I didn't feel any difference when playing around with this on the node that has run for a while

Reply:
Maybe you were just lucky not not happen to shutdown node during compaction 😄  Would rather turn it off just in case, as benefits are minor, but it adds risks of bad UX

Reply:
One more question. What is the benefit to enforce compaction on close?

Reply:
If we analyze the badger and what it stores for anything after the node is down. Just in general it's nicer to keep your db state always finalized.